### ISSUE TITLE
Requests need to merge headers

### ISSUE THREAD

**Issue by mxmauro** [algorand/py-algorand-sdk#169]
Reason: CF countermeasures returns 403 for any request with user agent containing the text `Python-urllib`.

Just try: `curl -v -A "Python-urllib" https://www.algorand.com/`

Regards.

**Comment by fabrice102** [comment]
For those affected by the issue, as a temporary workaround, you can manually change the User Agent.
E.g.,
```python
import algosdk.algod as algod
c = algod.AlgodClient('', 'https://api.algoexplorer.io', headers={'User-Agent': 'DoYouLoveMe?'})
print(c.status())
```

But hopefully CloudFlare will soon fix this issue.

**Comment by danmurphy1217** [comment]
> For those affected by the issue, as a temporary workaround, you can manually change the User Agent.
> E.g.,
> 
> ```python
> import algosdk.algod as algod
> c = algod.AlgodClient('', 'https://api.algoexplorer.io', headers={'User-Agent': 'DoYouLoveMe?'})
> print(c.status())
> ```
> 
> But hopefully CloudFlare will soon fix this issue.

Just an FYI that this does not completely solve the problem. It only serves as a solution if the Algorand SDK is not adding their own headers to HTTP requests later in the program. For example, I can define a user-agent early on as you suggest, but the `send_raw_transaction` function in **algosdk.v2client.algod** ALSO defines headers. Therefore, when I try to send a transaction I get a:

`TypeError: algosdk.v2client.algod.AlgodClient.algod_request() got multiple values for keyword argument 'headers'`

error since the initial headers that I set are passed as kwargs (and two arguments of the same name are not allowed).

I think this can and should be solved on the Algorand side, and I'm happy to write a patch.

**Comment by mxmauro** [comment]
@danmurphy1217  [this code](https://github.com/algorand/py-algorand-sdk/blob/master/algosdk/algod.py#L55) should merge the common headers added by the constructor with the ones on the `send_raw_transaction` call.

**Comment by danmurphy1217** [comment]
> @danmurphy1217 [this code](https://github.com/algorand/py-algorand-sdk/blob/master/algosdk/algod.py#L55) should merge the common headers added in the constructor with the ones on the `send_raw_transaction` call.

Let me provide you with some more context. I am calling the `client.send_transaction` method as follows:

```python
client.send_transaction(signed_transaction, headers={'User-Agent': '[value]'}
```

The `v2client.send_transaction` method in the Python SDK  returns the `send_raw_transaction` method, which looks like the following:

```python
    def send_raw_transaction(self, txn, **kwargs):
        """
        Broadcast a signed transaction to the network.
        Args:
            txn (str): transaction to send, encoded in base64
            request_header (dict, optional): additional header for request
        Returns:
            str: transaction ID
        """
        txn = base64.b64decode(txn)
        req = "/transactions"
        headers = { 'Content-Type': 'application/x-binary' }
        return self.algod_request("POST", req, data=txn, headers=headers, **kwargs)["txId"]
```
---


The root of the problem is on line 211 of **algosdk/v2client/algod.py**. The issue is that the following code:

```python
return self.algod_request("POST", req, data=txn, headers=headers, **kwargs)
```
passes the headers keyword argument and unpacks any kwargs as additional keyword arguments. Calling the `send_transaction` method with a `headers` argument earlier in the program means that this header value gets passed to `send_raw_transaction` as a kwarg. Ultimately, this means there are two `headers` args passed to the `self.algod_request` function, and this results in the following error:

```
TypeError: algosdk.v2client.algod.AlgodClient.algod_request() got multiple values for keyword argument 'headers'
```

My proposed solution is that we add a condition that checks if headers are already being passed as kwargs to `send_raw_transaction`. If they are, use the `Dict.update()` method to add the previously-passed headers to the ones defined in `send_raw_transaction`. We could also re-factor this into a decorator and apply it to several of the functions that already use this similar functionality.

**Comment by mxmauro** [comment]
To be honest, I'm not a Python expert. I think if one function does not propagates the headers as expected, can be fixed. If I pass `User-Agent` in the constructor or in every send tx call, I would expect to reach the destination.

**Comment by danmurphy1217** [comment]
@ian-algorand wouldn't mind getting someone at Algorand's perspective on this- what are your thoughts?

**Comment by jasonpaulos** [comment]
@danmurphy1217 your diagnosis above seems correct. We would gladly accept a PR which fixes this problem.

**Comment by danmurphy1217** [comment]
@jasonpaulos sounds good. I will tackle this tonight/tomorrow AM.

**Comment by jasonpaulos** [comment]
Fixed by #211
